src/celeritas/neutron/interactor/detail/CascadeCollider.hh

```diff
@@ -0,0 +1,235 @@
+ * implemented in collide and generateSCMfinalState methods. The cos\theta
+ * distribution in c.m. is inversely sampled using the tabulated cumulative
+ * distribution function (c.d.f) data in the kinetic energy and cos\theta bins
+ * which are implemented in G4CascadeFinalStateAlgorithm::GenerateTwoBody and
+ * G4NumIntTwoBodyAngDst::GetCosTheta methods.
+ */
+template<class Engine>
+CELER_FUNCTION auto CascadeCollider::operator()(Engine& rng) -> FinalState
+{
+    // Sample cos\theta of outgoing particles in the center of mass frame
+    real_type cdf = generate_canonical(rng);
+    TwodGridData cdf_grid = shared_.angular_cdf[ch_id_];
+    Grid energy_grid(cdf_grid.x, shared_.reals);
+    real_type cos_theta{0};
+
+    if (kin_energy_ < energy_grid.back())
+    {
+        // Find cos\theta from tabulated angular data for a given c.d.f.
+        Grid cos_grid(cdf_grid.y, shared_.reals);
+        TwodGridCalculator calc_cdf(cdf_grid, shared_.reals);
+
+        size_type idx = cos_grid.size() - 2;
+        real_type cdf_upper = 0;
+        real_type cdf_lower = 1;
+
+        do
+        {
+            cdf_upper = cdf_lower;
+            cdf_lower = calc_cdf({kin_energy_, cos_grid[idx]});
+        } while (cdf_lower > cdf && idx-- > 0);
+
+        real_type frac = (cdf - cdf_lower) / (cdf_upper - cdf_lower);
+        cos_theta = fma(frac, cos_grid[idx + 1] - cos_grid[idx], cos_grid[idx]);
+    }
+    else
+    {
+        // Sample the angle outside tabulated data (unlikely)
+        real_type slope = 2 * energy_grid.back() * ipow<2>(cm_p_);
+        cos_theta = std::log(1 + cdf * std::expm1(2 * slope)) / slope - 1;
+    }
+
+    // Sample the momentum of outgoing particles in the center of mass frame
+    Real3 mom = cm_p_ * from_spherical(cos_theta, sample_phi_(rng));
+
+    // Rotate the momentum along the reference z-axis
+    FourVector fv = {mom,
+                     std::sqrt(dot_product(mom, mom)
+                               + ipow<2>(value_as<MevMass>(bullet_.mass)))};
+
+    // Find the final state of outgoing particles
+    FinalState result = {bullet_, target_};
+
+    FourVector cm_momentum = target_.four_vec;
+    boost(-cm_velocity_, &cm_momentum);
+
+    Real3 cm_dir = make_unit_vector(-cm_momentum.mom);
+    real_type vel_parallel = dot_product(cm_velocity_, cm_dir);
+
+    // Degenerated if velocity perpendicular to the c.m. momentum is small
+    bool degenerated
+        = (dot_product(cm_velocity_, cm_velocity_) - ipow<2>(vel_parallel)
+           < this->epsilon());
+
+    if (!degenerated)
+    {
+        Real3 vscm = make_unit_vector(cm_velocity_ - vel_parallel * cm_dir);
+        Real3 vxcm = make_unit_vector(cross_product(cm_dir, cm_velocity_));
+        if (norm(vscm) > this->epsilon() && norm(vxcm) > this->epsilon())
+        {
+            result[0].four_vec
+                = {{fv.mom[0] * vscm + fv.mom[1] * vxcm + fv.mom[2] * cm_dir},
+                   fv.energy};
```

Comment:
This is exactly why I didn't want to make "array operators": even though this reads like a simple algebra expression, C++ is creating an insane number of temporaries here. I created [some test kernels](https://github.com/celeritas-project/cuda-test-snippets/tree/main/celer-vec-add) replicating this operation: as written, this takes 35 registers and 280 (!) bytes of stack space. Rewriting inside a `for` loop reduces the usage to 32 registers and 88 bytes of stack space, or even less (depending on how it's written).

```suggestion
                for (int i = 0; i < 3; ++i)
                {
                    result[0].four_vec[i] = fv.mom[0] * vscm[i]
                                            + fv.mom[1] * vxcm[i]
                                            + fv.mom[2] * cm_dir[i];
                }
                result[0].four_vec.energy = fv.energy;
```

Reply:
Good to know!  I thought that they are very much equivalent. Anyway, changed as suggested.

Reply:
Yeah, unfortunately C++ forces the creation of temporaries that all store local copies of the intermediate compilation. The only ways to get around it are to manually unroll like this or to use a complicated template metaprogramming math library like Eigen or Blitz++ (neither of which are CUDA compatible).